### PR TITLE
Feature/clover reconstruct

### PR DESCRIPTION
**Pull request by maddyscientist**
Feature-add branch
* Add clover-reconstruct CMake option `QUDA_CLOVER_RECONSTRUCT` to QUDA (enabled by default) which compresses the clover field when using the native field order.  This is only supported when `QUDA_CLOVER_DYNAMIC_INVERSE=ON`
* This reduces the memory footprint and improves performance for all kernels that utilize clover fields
* Clean up of handling when dynamic-inverse is enabled to remove hacks in the interface
* Dynamic-clover now utilizes double precision for the Cholesky for double and single precision fields, and single precision for half and quarter precision fields.  This improves solver stability.
* Unified `cpuCloverField` and `cudaCloverField`: we now only have `CloverField`
* Add support for `FLOAT8` ordering for half and quarter precision clover fields
* Utilize global norm scale for fixed-point clover field instead of per-site normalization
* Fixes to tunable_reduction.h with respect to correct setting of the `aux` key
* Remove legacy `writeAuxString` method

### CODE REVIEW

**Comment by weinbe2** [comment]
> Dynamic-clover now utilizes double precision for the Cholesky for double and single precision fields

While I understand the sentiment of stability, I'm not sure I like the "silent fp64" in the single precision operator.

**Comment by weinbe2** [comment]
> > Dynamic-clover now utilizes double precision for the Cholesky for double and single precision fields
> 
> While I understand the sentiment of stability, I'm not sure I like the "silent fp64" in the single precision operator.

Conversation resolved offline. While I still don't necessary like the "silent fp64" in a single precision operator, there's still an option to compile w/out dynamic clover -> the inverse clover gets precomputed and directly applied in fp32, giving a way out if avoiding fp64 is truly desired.

**Comment by weinbe2** [comment]
I've gone through a full cursory review and I'm very impressed with this PR. I'm about to dig into some hands-on testing

one request I have in the interim is for some documentation on these new features on the wiki; some type of "clover fermion best practices". This could go under `QUDA Solver Acceleration Strategies` (similar to documentation for advanced dwf features, etc) or `Using QUDA in your LQCD applications`, or wherever else is deemed fit.

File: include/clover_field.h [hunk 1]
```diff
@@ -32,12 +79,10 @@ namespace quda {
   };
 
   struct CloverFieldParam : public LatticeFieldParam {
-    bool direct; // whether to create the direct clover 
-    bool inverse; // whether to create the inverse clover
+    bool reconstruct; /** Whether to create a compressed that requires reconstruction */
```
Comment (cpviolator) [inline include/clover_field.h]: "a compressed clover field that"?

File: include/clover_field_order.h [hunk 2]
```diff
@@ -519,49 +558,60 @@ namespace quda {
...
+                bool enable_reconstruct_ = clover::reconstruct(), bool huge_alloc = false>
+      struct FloatNOrder {
+        static constexpr bool enable_reconstruct = enable_reconstruct_;
+        using Accessor = FloatNOrder<Float, length, N, add_rho, enable_reconstruct, huge_alloc>;
+        using real = typename mapper<Float>::type;
+        typedef typename VectorType<Float, N>::type Vector;
+        typedef typename AllocType<huge_alloc>::type AllocInt;
+        typedef float norm_type;
+        static constexpr int Ns = 4;
+        static constexpr int Nc = 3;
```
Comment (cpviolator) [inline include/clover_field_order.h:571]: Can you place a static_assert here on 3?

File: include/enum_quda_fortran.h [hunk 3]
```diff
@@ -237,9 +237,10 @@
 #define QUDA_FLOAT_CLOVER_ORDER 1   // even-odd float ordering 
 #define QUDA_FLOAT2_CLOVER_ORDER 2   // even-odd float2 ordering
 #define QUDA_FLOAT4_CLOVER_ORDER 4   // even-odd float4 ordering
-#define QUDA_PACKED_CLOVER_ORDER 5    // even-odd packed
-#define QUDA_QDPJIT_CLOVER_ORDER 6 // lexicographical order packed
-#define QUDA_BQCD_CLOVER_ORDER 7 // BQCD order which is a packed super-diagonal form
+#define QUDA_FLOAT8_CLOVER_ORDER 8   // even-odd float4 ordering
```
Comment (cpviolator) [inline include/enum_quda_fortran.h]: "// even-odd float8 ordering"

File: include/enum_quda.h [hunk 4]
```diff
@@ -254,6 +254,7 @@ typedef enum QudaCloverFieldOrder_s {
   QUDA_FLOAT_CLOVER_ORDER = 1,  // even-odd float ordering
   QUDA_FLOAT2_CLOVER_ORDER = 2, // even-odd float2 ordering
   QUDA_FLOAT4_CLOVER_ORDER = 4, // even-odd float4 ordering
+  QUDA_FLOAT8_CLOVER_ORDER = 8, // even-odd float4 ordering
```
Comment (cpviolator) [inline include/enum_quda.h]: "// even-odd float8 ordering"

File: include/clover_field_order.h [hunk 5]
```diff
@@ -881,20 +1026,36 @@ namespace quda {
...
   // single precision uses Float4
-  template<int N, bool add_rho> struct clover_mapper<float,N,add_rho> { typedef clover::FloatNOrder<float, N, 4, add_rho> type; };
+  template <int N, bool add_rho, bool enable_reconstruct> struct clover_mapper<float, N, add_rho, enable_reconstruct> {
+    using type = clover::FloatNOrder<float, N, 4, add_rho, enable_reconstruct>;
+  };
+
+#ifdef FLOAT8
+#define N8 8
+#else
+#define N8 4
```
Comment (cpviolator) [inline include/clover_field_order.h]: Can this be changed to something like

```C++
#ifdef FLOAT8
 #define FLOATN 8
 #else
 #define FLOATN 4
```
with the appropriate substitutions to `N8` below?

**Review by cpviolator — CHANGES_REQUESTED** [review]
All looks good to me. There some instances of Nc=3 specific code, but I can see most of them from the hardcoded values of 6, 30, 72, etc, so not too hard to upgrade the loops in the Nc agnostic branch.

File: include/clover_field_order.h (hunk 5 above)
Comment (maddyscientist) [reply include/clover_field_order.h]: 84359ab (and I applied the same cleanup to gauge_field_order.h and color_spinor_field_order.h)

File: include/enum_quda.h (hunk 4 above)
Comment (maddyscientist) [reply include/enum_quda.h]: 84359ab

File: include/enum_quda_fortran.h (hunk 3 above)
Comment (maddyscientist) [reply include/enum_quda_fortran.h]: 84359ab

File: include/clover_field_order.h (hunk 2 above)
Comment (maddyscientist) [reply include/clover_field_order.h:571]: 84359ab

File: include/clover_field.h (hunk 1 above)
Comment (maddyscientist) [reply include/clover_field.h]: 84359ab

**Comment by maddyscientist** [comment]
Thanks @cpviolator for the review.  I've addressed all your comments in 84359ab.

Also fixed a nasty UB bug with MPI broadcasts getting mixed up in d71d9d0.  PR revalidated, both single and multi-GPU.

File: include/instantiate.h [hunk 6]
```diff
@@ -151,6 +151,31 @@ namespace quda
     }
   }
 
+  /**
+     @brief This instantiate2 function is used to instantiate the
+     precisions, with double precision always enabled.  This is a
+     temporary addition until we fuse this with the original function
+     above when we enforce C++17
```
Comment (mathiaswagner) [inline include/instantiate.h:158]: What is stopping us from just enforcing C++17 now?

File: CMakeLists.txt [hunk 7]
```diff
@@ -474,6 +476,10 @@ if(QUDA_HETEROGENEOUS_ATOMIC AND QUDA_JITIFY)
   message(SEND_ERROR "QUDA_HETEROGENEOUS_ATOMIC=ON does not support JITIFY)")
 endif()
 
+if(QUDA_CLOVER_RECONSTRUCT AND NOT QUDA_CLOVER_DYNAMIC)
```
Comment (mathiaswagner) [inline CMakeLists.txt:479]: thoughts on replacing this if by using
```
cmake_dependent_option(QUDA_CLOVER_RECONSTRUCT "set to ON to enable compressed clover storage" OFF "QUDA_CLOVER_DYNAMIC" OFF)
```
?
```

File: include/instantiate.h (hunk 6 above)
Comment (maddyscientist) [reply include/instantiate.h:158]: Nothing stopping us.  I would probably prefer C++17 related cleanup in a separate PR in favor of getting this merged first.

File: CMakeLists.txt (hunk 7 above)
Comment (maddyscientist) [reply CMakeLists.txt:479]: No strong thoughts.  Personally
I don’t like options disappearing from `ccmake` when certain options are set, and I like to see them, even if they error when enabled / disabled.

File: include/instantiate.h (hunk 6 above)
Comment (weinbe2) [reply include/instantiate.h:158]: > Nothing stopping us. I would probably prefer C++17 related cleanup in a separate PR in favor of getting this merged first.

+1 on making it separate

File: CMakeLists.txt (hunk 7 above)
Comment (mathiaswagner) [reply CMakeLists.txt:479]: works for me. I was just annoyed that to disable dynamic I had to switch to advanced to disable reconstruct separately.

**Comment by maddyscientist** [comment]
Noting that I've applied some cleanup to the `clover::FloatNOrder::load` and `clover::FloatNOrder::save` functions to reduce code duplication and improve readability.  As suggested above, I've also made a wiki page to discuss some of the clover fermion options: https://github.com/lattice/quda/wiki/Clover-Fermions.

**Comment by weinbe2** [comment]
Making sure I understand what I need to test the outer product of:
* FLOAT8 on/off
* (dynamic off/reconstruct off) ; (dynamic on/reconstruct off) ; (dynamic on/reconstruct on)
Which is 6 combinations total. Am I missing any combination?

Also---do these different combinations get appropriately distinguished in tunekeys?

**Comment by maddyscientist** [comment]
@weinbe2 your combinations look good.

Regarding whether these combinations are distinguished with tunekeys: in keeping in general with other compile-time options as opposed to run-time options, there is no effort to break degeneracy

**Review by weinbe2 — APPROVED** [review]
For the above six combinations of compile flags, I ran:
* A full `ctest`
* A separate test of multigrid across precisions, preconditionings, clover/twisted-clover, and Nc

I did not test different partitions because the clover field is an inherently rank-local quantity.

Everything passed and looked good, so this gets my approve!

**Review by cpviolator — APPROVED** [review]
All good! Awesome PR @maddyscientist

**Pull request merged**